acinclude/compiler-flags.m4

```diff
@@ -113,18 +113,44 @@ AC_DEFUN([SQUID_CC_GUESS_VARIANT], [
   ]) dnl AC_CACHE_CHECK
  ]) dnl AC_DEFUN
 
-AC_DEFUN([SQUID_CC_ADD_CXXFLAG_IF_SUPPORTED_INTERNAL],[
+dnl SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED helper
+dnl $1 is a compiler warning option (e.g., -Wall).
+dnl $2 is a "squid_cv_cc_arg<snake_case_warning_name_equivalent>" string.
+AC_DEFUN([SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED_],[
+  AC_REQUIRE([SQUID_CC_GUESS_VARIANT])
   SQUID_STATE_SAVE([CXXARGTEST])
   CXXFLAGS="$CXXFLAGS $SQUID_CXXFLAGS"
-  SQUID_CC_CHECK_ARGUMENT([$2],[$1])
+  case "$squid_cv_compiler" in
+    gcc)
+      # Testing with -Werror -Wfoobar does not work well because GCC ignores
+      # unsupported _negative_ options, so we test with -Werror=foobar instead
+      # (where "foobar" is a name of a warning that may be given to us in
+      # positive -Wfoobar or negative -Wno-foobar form).
+      SQUID_CC_CHECK_ARGUMENT([$2],m4_bpatsubst([$1],[^-W], [-Werror=]))
+      ;;
+    clang)
+      # Testing with -Werror=foobar (as we do for GCC above) is useless
+      # because clang does not recognize that pattern as a -Werror
+      # specialization, so we test with -Werror -Wfoobar instead.
+      SQUID_CC_CHECK_ARGUMENT([$2],[-Werror $1])
+      ;;
+    *)
+      # We lack code to reliably test whether this compiler supports a given
+      # warning. Some compilers (e.g, icc) succeed with bogus warning names.
+      # If $squid_cv_cxx_option_werror is set, we add that option because it
+      # helps in some (but not all) known cases.
+      SQUID_CC_CHECK_ARGUMENT([$2],[$squid_cv_cxx_option_werror $1])
+      ;;
+  esac
   SQUID_STATE_ROLLBACK([CXXARGTEST])
   AS_IF([test "x${$2}" = "xyes"],[SQUID_CXXFLAGS="$SQUID_CXXFLAGS $1"])
 ])
 
-dnl argument is a compiler flag. It will be attempted, and if suppported
-dnl it will be added to SQUID_CXXFLAGS in the same order as calls to the macro
-AC_DEFUN([SQUID_CC_ADD_CXXFLAG_IF_SUPPORTED],[
-  SQUID_CC_ADD_CXXFLAG_IF_SUPPORTED_INTERNAL($1,m4_bpatsubst(m4_tolower([squid_cv_cc_arg$1]),[[^a-zA-Z0-9_]], [_]))
+dnl The argument is a compiler warning option (e.g. -Wall). If linking a
+dnl warning-free program while using the given warning succeeds, then the
+dnl option is added to SQUID_CXXFLAGS in the same order as calls to the macro.
+AC_DEFUN([SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED],[
+  SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED_($1,m4_bpatsubst(m4_tolower([squid_cv_cc_arg$1]),[[^a-zA-Z0-9_]], [_]))
```

Comment:
> Whats with the rename for these macros?

These are completely different macros: Removed macros were for all compiler flags. New macros are for compiler warnings only; they are not meant for other compiler flags/options. This scope change is important enough to be reflected in the macro name IMO.

FWIW, I am also unhappy with the `_SUPPORTED` part which promises more than these macros can deliver. The macro description no longer makes that promise, but I also wanted to use something like `_SEEMS_SUPPORTED` or `_APPEARS_TO_WORK` in the macro name. I did not do it because I was worried abut the macro names hitting some (unknown to me) length limit and getting into costly debates about names. Better names are always welcome, of course.